src/main/java/com/pingcap/tikv/PDClient.java

```diff
@@ -229,60 +224,65 @@ void close() {
     }
   }
 
-  public GetMembersResponse getMembers() {
-    List<HostAndPort> pdAddrs = getConf().getPdAddrs();
-    checkArgument(pdAddrs.size() > 0, "No PD address specified.");
-    for (HostAndPort url : pdAddrs) {
-      try {
-        ManagedChannel probChan = getChannel(url.getHostText() + ":" + url.getPort());
-        PDGrpc.PDBlockingStub stub = PDGrpc.newBlockingStub(probChan);
-        GetMembersRequest request =
-            GetMembersRequest.newBuilder().setHeader(RequestHeader.getDefaultInstance()).build();
-        return stub.getMembers(request);
-      } catch (Exception ignore) {}
+  public GetMembersResponse getMembers(HostAndPort url) {
+    try {
+      ManagedChannel probChan = getChannel(url.getHostText() + ":" + url.getPort());
+      PDGrpc.PDBlockingStub stub = PDGrpc.newBlockingStub(probChan);
+      GetMembersRequest request =
+          GetMembersRequest.newBuilder().setHeader(RequestHeader.getDefaultInstance()).build();
+      return stub.getMembers(request);
+    } catch (Exception e) {
+      logger.warn("failed to get member from pd server.", e);
     }
     return null;
   }
 
-  public void updateLeader(GetMembersResponse resp) {
-    String leaderUrlStr = "URL Not Set";
-    try {
-      long ts = System.nanoTime();
-      synchronized (this) {
-        // Lock for not flooding during pd error
-        if (leaderWrapper != null && leaderWrapper.getCreateTime() > ts) return;
-
-        if (resp == null) {
-          resp = getMembers();
-          if (resp == null) return;
-        }
-        Member leader = resp.getLeader();
-        List<String> leaderUrls = leader.getClientUrlsList();
-        if (leaderUrls.isEmpty()) return;
-        leaderUrlStr = leaderUrls.get(0);
-        URL tURL = new URL(leaderUrlStr);
-        HostAndPort newLeader = HostAndPort.fromParts(tURL.getHost(), tURL.getPort());
-        leaderUrlStr = newLeader.toString();
-        // TODO: Why not strip protocol info on server side since grpc does not need it
-        if (leaderWrapper != null && leaderUrlStr.equals(leaderWrapper.getLeaderInfo())) {
-          return;
-        }
+  private void switchLeader(List<String> leaderURLs) {
```

Comment:
You likely need lock or sync here. Consider if it's a heavy PD search case, eg, get / put. Each  call will consult PD for a region. Once leader down and switched, you have tons of threads trying to do a leader switch and DDOS new PD cluster.

Reply:
PD switch is part of error handling case. Even a whole sync block will be fine since normal code path will not pay the penalty.

Reply:
FIXED